AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/Model+Compare.swift

```diff
@@ -0,0 +1,142 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+// swiftlint:disable cyclomatic_complexity
+// swiftlint:disable syntactic_sugar
+extension ModelSchema {
+
+    /// Compare two `Model` based on a given `ModelSchema`
+    /// Returns true if equal, false otherwise
+    public func isEqual(_ model1: Model, _ model2: Model) -> Bool {
+        for (fieldName, modelField) in fields {
+            // read only fields are skipped for eqaulity check as they are created by the service
+            if modelField.isReadOnly {
+                continue
+            }
+
+            let value1 = model1[fieldName] ?? nil
+            let value2 = model2[fieldName] ?? nil
+
+            // check equality for different `ModelFieldType`
+            switch modelField.type {
+            case .string:
+                guard let value1Optional = value1 as? String?, let value2Optional = value2 as? String? else {
+                    return false
+                }
+                if let value1 = value1Optional, let value2 = value2Optional, value1 != value2 {
+                    return false
+                } else {
+                    continue
```

Comment:
if `value1Optional` cannot be unwrapped because it's `nil` but value2Optional could have been unwrapped, then we should return false, right? since one side is nil and the other side is valid, but this logic would continue and move onto the next field